crates/lang/src/env_access.rs

```diff
@@ -831,10 +830,17 @@ where
         self,
         signature: &[u8; 65],
         message_hash: &[u8; 32],
-    ) -> Result<ECDSAPublicKey> {
+    ) -> Result<[u8; 33]> {
         let mut output = [0; 33];
         ink_env::ecdsa_recover(signature, message_hash, &mut output)
-            .map(|_| output.into())
+            .map(|_| output)
+            .map_err(|_| Error::EcdsaRecoveryFailed)
+    }
+
+    pub fn ecdsa_to_eth_address(self, pubkey: &[u8; 33]) -> Result<[u8; 20]> {
```

Comment:
I'd be curious of LLVM is able to do copy elision of the return value here. It is not guaranteed by Rust (yet).